### PR TITLE
feat: opt-in pipeline fit optimisations, native bucketize, and sampled fitting support with caching

### PR DESCRIPTION
**Pull request by ConorWorthington**
### Description
Provide a short description of the PR changes.

The below checklists come from the docs page on adding new transformers [here](../blob/main/docs/adding_transformer.md)
### Keras Layer Checklist
Verify that:
- [ ] The new Keras layer extends [BaseLayer](../blob/main/src/kamae/tensorflow/layers/base.py)
- [ ] The `_call` method has been implemented in the new layer.
- [ ] The `compatible_dtypes` property is defined in the new layer.
- [ ] The new layer is decorated with `@tf.keras.utils.register_keras_serializable(package=kamae.__name__)`.
- [ ] The new layer takes a `name`, `input_dtype`, and `output_dtype` as arguments to the constructor and that this is passed to the super constructor.
- [ ] The Keras layer is serializable. I have implemented the `get_config` method.
- [ ] There are unit tests of the new layer. 
- [ ] There is a specific test of layer serialisation added [here](../blob/main/tests/kamae/tensorflow/test_layer_serialisation.py).
- [ ] The new layer is imported in the [__init__.py](../blob/main/src/kamae/tensorflow/layers/__init__.py) file in the `layers` directory.

### Spark Transformer/Estimator Checklist
Verify that:
- [ ] The new Spark Transformer extends [BaseTransformer](../blob/main/src/kamae/spark/transformers/base.py).
- [ ] If the new transform needs a fit method, a Spark Estimator has been implemented that extends [BaseEstimator](../blob/main/src/kamae/spark/estimators/base.py).
- [ ] The instructions in the above docs page have been followed for the `__init__` and `setParams` methods.
- [ ] The transformer uses one of the input/output mixin classes from [base.py](../blob/main/src/kamae/spark/params/base.py).
- [ ] If the new transformer requires more parameters that would need to be serialised to the Spark ML pipeline, there is a implemented parameter class by extending the `Params` class [here](https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.ml.param.Params.html).
- [ ] The `compatible_dtypes` property has been implemented to specify the input/output data types that my transformer/estimator supports.
- [ ] A Keras subclassed layer is returned in the transformer's `get_tf_layer` method.
- [ ] There are unit tests of the new transform. In particular, there are parity tests between the Spark and Keras implementations.
- [ ] The new transformer/estimator is imported in the [__init__.py](../blob/main/src/kamae/spark/transformers/__init__.py) file in the `transformers`/`estimators` directory.

Finally, please verify that:
- [ ] There is a new entry (alphabetical order) in the README table describing the new layer/transformer

### CODE REVIEW

File: src/kamae/spark/pipeline/pipeline.py [hunk 1]
```diff
@@ -38,17 +39,69 @@ class KamaeSparkPipeline(Pipeline):
...
+        :param cacheIntermediateData: If True, persist the working DataFrame at
+        each estimator-fit boundary to avoid re-scanning the upstream lineage.
+        False (default) disables it.
+        :param pruneInputColumns: If True, drop input columns no stage consumes
+        before fitting. False (default) disables it.
         :returns: None - class instantiated.
         """
-        super().__init__(stages=stages)
+        kwargs = self._input_kwargs
+        super().__init__()
```
Comment (georyetti) [inline src/kamae/spark/pipeline/pipeline.py]: The new super call drops the `stages=stages`, was this intended?

File: src/kamae/spark/pipeline/pipeline.py [hunk 2]
```diff
@@ -38,17 +39,69 @@ class KamaeSparkPipeline(Pipeline):
...
+        self._setDefault(
+            checkpointInterval=0,
```
Comment (georyetti) [inline src/kamae/spark/pipeline/pipeline.py]: Default to `None` instead of 0?

File: src/kamae/spark/pipeline/pipeline.py [hunk 3]
```diff
@@ -132,25 +249,120 @@ def collect_estimator_parents(
...
+                    continue
+                if not stage.isDefined(param):
+                    continue
+                value = stage.getOrDefault(param)
+                if isinstance(value, str):
+                    required_input_columns.add(value)
+                elif isinstance(value, (list, tuple)):
+                    required_input_columns.update(
+                        item for item in value if isinstance(item, str)
+                    )
```
Comment (georyetti) [inline src/kamae/spark/pipeline/pipeline.py:424]: This feels more complex than I think it needs to be. Two things:
1. Why do we need to iterate through the stage.params at all if we have already added the inputs using `get_layer_inputs_outputs`
2. Even if we do need, we can just check for presence of `inputCol` or `inputCols` as a stage always has strictly one of these defined. `if stage.hasParam("inputCol") and stage.isDefined("inputCol"): value = stage.getInputCol()` and after check input cols. But this is really what `get_layer_inputs_outputs` does.

File: src/kamae/spark/pipeline/pipeline.py [hunk 4]
```diff
@@ -132,25 +249,120 @@ def collect_estimator_parents(
...
+
+        Returned unchanged if there are no unused columns to drop.
+
+        :param dataset: Input DataFrame to prune.
+        :param stages: Expanded pipeline stages.
+        :returns: DataFrame projected to the columns the pipeline consumes.
+        """
+        required_input_columns = self.collect_required_input_columns(stages)
+        columns_to_keep = [c for c in dataset.columns if c in required_input_columns]
+        if columns_to_keep and len(columns_to_keep) < len(dataset.columns):
```
Comment (georyetti) [inline src/kamae/spark/pipeline/pipeline.py]: Pedantic but do we need that second condition? If we are inside this function then we are pruning, and columns_to_keep is always a subset. So I would just check it's not empty and otherwise select

File: src/kamae/spark/pipeline/pipeline.py [hunk 5]
```diff
@@ -67,15 +120,79 @@ def getStages(self) -> List["KamaePipelineStage"]:
...
 
+    def setCheckpointInterval(self, value: int) -> "KamaeSparkPipeline":
+        """
+        Sets the `checkpointInterval` parameter.
+
+        :param value: Number of stages between reliable checkpoint calls during
+        fit. 0 (or None) disables checkpointing.
+        :returns: KamaeSparkPipeline object with checkpointInterval set.
+        """
+        return self._set(checkpointInterval=value)
```
Comment (georyetti) [inline src/kamae/spark/pipeline/pipeline.py:198]: We should error on checkpoint interval being negative here. Personally I think we should error on 0 too and treat None as the no checkpoint behaviour

File: src/kamae/spark/pipeline/pipeline.py [hunk 6]
```diff
@@ -67,15 +120,79 @@ def getStages(self) -> List["KamaePipelineStage"]:
...
+        :param checkpointInterval: Number of stages between reliable
+        checkpoint(eager=True) calls during fit. 0 (default) disables it.
+        :param cacheIntermediateData: If True, persist the working DataFrame at
+        each estimator-fit boundary. False (default) disables it.
+        :param pruneInputColumns: If True, drop input columns no stage consumes
+        before fitting. False (default) disables it.
+        :returns: KamaeSparkPipeline object with params set.
         """
         kwargs = self._input_kwargs
         return self._set(**kwargs)
```
Comment (georyetti) [inline src/kamae/spark/pipeline/pipeline.py]: This set params does not use the setter methods at all. So any validation in them will not be respected when the user passes arguments to the __init__ as opposed to using the setter method. If you check how I defined the setParams for the estimator and transformer you can see I use the setter method.

File: src/kamae/spark/transformers/bucketize.py [hunk 7]
```diff
@@ -141,31 +140,34 @@ def _transform(self, dataset: DataFrame) -> DataFrame:
...
         splits = self.getSplits()
-        # We need to create a UDF to perform binary search on the splits.
 
-        def bucketize(value: Optional[Union[float, int]]) -> Optional[int]:
-            # If null, keep null. There is no best bucket to place these into.
-            if value is None:
-                return None
-            # We add 1 because we want to reserve the 0 index for mask/padding.
-            return bisect_right(splits, value) + 1
+        def bucketize(value: Column) -> Column:
```
Comment (georyetti) [inline src/kamae/spark/transformers/bucketize.py:144]: Same to assume that all this bucket logic changes to this transformer are not meant to be in this PR?

File: pyproject.toml [hunk 8]
```diff
@@ -12,6 +12,7 @@ requires-python = ">=3.10,<3.13"
 dependencies = [
     "pyspark>=3.4.0,<4.0.0",
     "pandas>=1.3.4,<3.0.0",
+    "pyarrow>=4.0.0",
```
Comment (georyetti) [inline pyproject.toml:15]: Adding new dependencies needs a uv lock pls

**Review by georyetti — CHANGES_REQUESTED** [review]
Mostly minor changes to pipeline logic. Also bucketize changes I assume should not be here. Lastly you need to run a uv lock.

File: src/kamae/spark/pipeline/pipeline.py (hunk 5 above)
Comment (ConorWorthington) [reply src/kamae/spark/pipeline/pipeline.py:198]: Updated

File: src/kamae/spark/transformers/bucketize.py (hunk 7 above)
Comment (ConorWorthington) [reply src/kamae/spark/transformers/bucketize.py:144]: I was looking for inefficiencies in the library... I was going to patch more but only bucketize was really impacted in a way I could speed up simply. We don't really use it but figured it was a good improvement so may as well include.

File: pyproject.toml (hunk 8 above)
Comment (ConorWorthington) [reply pyproject.toml:15]: Lock updated

File: src/kamae/spark/pipeline/pipeline.py (hunk 3 above)
Comment (ConorWorthington) [reply src/kamae/spark/pipeline/pipeline.py:424]: You're right that inputCol/inputCols are already covered by get_layer_inputs_outputs but the sweep isn't for those.

It's there for the other column params a stage reads that aren't its input col: maskCols/relevanceCol on ConditionalStandardScaleEstimator, and queryIdCol on the listwise transformers. Those get read at fit time, so if we don't keep them, pruneInputColumns=True drops them and the fit blows up with "column not found". There are regression tests covering exactly that.

If the suffix-matching feels too hacky, I'm happy to swap it for a small get_fit_input_columns() hook on the handful of stages that need it instead. Let me know what you'd prefer.

Comment (georyetti) [reply src/kamae/spark/pipeline/pipeline.py:424]: Ok makes sense.

File: src/kamae/spark/transformers/bucketize.py (hunk 7 above)
Comment (georyetti) [reply src/kamae/spark/transformers/bucketize.py:144]: Hmmm ok if you want to keep it I will have more comments, will add them then you can decide.

File: src/kamae/spark/transformers/bucketize.py [hunk 9]
```diff
@@ -141,31 +140,34 @@ def _transform(self, dataset: DataFrame) -> DataFrame:
...
+            func=bucketize,
         )
 
         return dataset.withColumn(
             self.getOutputCol(),
             output_col,
         )
 
-    def get_keras_layer(self) -> tf.keras.layers.Layer:
+    def get_keras_layer(self) -> keras.layers.Layer:
```
Comment (georyetti) [inline src/kamae/spark/transformers/bucketize.py]: Why are we making this type hint as keras layer? The layer still uses tf only functions so we should revert this back

**Review by georyetti — CHANGES_REQUESTED** [review]
Effectively fine, just revert the type hint for bucketize back to tf layer as its still tf only

**Review by georyetti — DISMISSED** [review]

File: src/kamae/spark/pipeline/pipeline.py [hunk 10]
```diff
@@ -426,6 +504,8 @@ def _fit(self, dataset: DataFrame) -> "KamaeSparkPipelineModel":
                     dataset = model.transform(dataset)
         if cached_dataset is not None:
             cached_dataset.unpersist()
+        if estimator_input_cache is not None:
+            estimator_input_cache.unpersist()
```
Comment (georyetti) [inline src/kamae/spark/pipeline/pipeline.py]: Got a little confused here as we have 3 dataframes:
- dataset
- cached_dataset
- new_cached

And we set the first 2 to the 3rd, but then only unpersist the 2nd...?

Can we simplify this to just persist `dataset`?

File: src/kamae/spark/pipeline/pipeline.py [hunk 11]
```diff
@@ -404,6 +466,22 @@ def _fit(self, dataset: DataFrame) -> "KamaeSparkPipelineModel":
...
+                    estimator_input_cached = True
+                    live_columns = self.collect_required_input_columns(
+                        expanded_pipeline_stages[index:]
+                    )
+                    keep_columns = [c for c in dataset.columns if c in live_columns]
+                    if keep_columns and len(keep_columns) < len(dataset.columns):
+                        estimator_input_cache = dataset.select(*keep_columns).persist(
+                            StorageLevel.MEMORY_AND_DISK
+                        )
+                        dataset = estimator_input_cache
```
Comment (georyetti) [inline src/kamae/spark/pipeline/pipeline.py]: Similar here, what is the use of estimator_input_cache here if we just set to dataset and then have to remember to unpersist()? Can we not just reuse the dataset variable and unpersist() just that at the end?

**Review by georyetti — CHANGES_REQUESTED** [review]
Sorry some minor comments, I don't think its particularly blocking so can approve if you have good reason for it.

**Comment by ConorWorthington** [comment]
Pushed a new commit (`2e0a0cd`) with two things:

**1. New opt-in param: `fitSampleFraction` (+ `fitSampleSeed`)**

This is a fit-time optimisation for the case that motivated the caching work: a pipeline of several independent estimators (each `ArrayConcatenate -> ConditionalStandardScaleEstimator` sampling internally at ~0.005) where every estimator does its own full Bernoulli scan of the wide source. Persisting the intermediate/estimator-input frame doesn't help there because the frame is too big for the storage-memory budget — it spills (~1.7 TiB) and the executors GC-thrash.

Insight: the fit only needs a *sample*. So when `fitSampleFraction` is set to a float in `(0, 1]`, `_fit`:
- draws a single `dataset.sample(fraction, seed=fitSampleSeed)` up-front, `persist(MEMORY_AND_DISK)`s it and forces one `count()` so it materialises exactly once;
- runs the whole fit loop against that shared sample;
- temporarily forces each estimator's own `sampleFraction` to `None` (restored in a `finally`) so we don't sample-a-sample;
- unpersists the sample in the `finally`.

The returned model is unchanged — transformers still apply to the full dataset at transform time; only fit-time stats come from the sample.

**Caveats (also in the param docstring + a runtime warning):**
- It's incompatible with `cacheIntermediateData`/`cacheEstimatorInput` (they persist the frames this option avoids), so enabling it warns and disables them.
- It fits *all* estimators from one shared sample, so it's only correct for sample-robust statistics (mean/std/quantiles, e.g. `ConditionalStandardScale`). Vocabulary builders (StringIndexer/OneHot), min/max scalers and distinct counts need exact/global stats and will be inaccurate — hence the runtime warning.

Tests cover: `None` = unchanged; sampled vs full mean/stddev within a loose tolerance on seeded data; an accumulator assertion that the source is scanned once (not once-per-estimator); and the mutual-exclusion warning.

Note: this is correctness-tested, but I have **not** benchmarked the actual speedup at prod scale — the single-scan behaviour is proven, the wall-clock win is expected but unmeasured.

**2. Addressed your two latest comments on the cache bookkeeping**

Since `cacheIntermediateData` and `cacheEstimatorInput` are mutually exclusive, I collapsed the `cached_dataset` / `estimator_input_cache` / `new_cached` handles into a single `persisted_frame` with one `unpersist()`. I kept it as a separate handle (rather than just unpersisting `dataset` at the end) because `dataset` gets reassigned by `model.transform(...)` between boundaries, so by loop-end it no longer points at the persisted frame — there's a docstring note explaining that.

File: src/kamae/spark/pipeline/pipeline.py [hunk 12]
```diff
@@ -764,3 +849,43 @@ class KamaeSparkPipelineWriter(PipelineWriter):
...
+            "language": "Python",
+        }
+        # Only the explicitly-set, non-stages params; defaults stay implicit so
+        # older saves (without this metadata) still load with correct defaults.
+        pipeline_params = {
+            p.name: self.instance.getOrDefault(p)
+            for p in self.instance.params
+            if p.name != "stages" and self.instance.isSet(p)
+        }
+        DefaultParamsWriter.saveMetadata(
```
Comment (georyetti) [inline src/kamae/spark/pipeline/pipeline.py]: Can we use the kamae classes we have here: `KamaeDefaultParamsWriter`

File: src/kamae/spark/pipeline/pipeline.py [hunk 13]
```diff
@@ -754,7 +831,15 @@ def load(self, path: str) -> KamaeSparkPipeline:
         """
         metadata = DefaultParamsReader.loadMetadata(path, self.sc)
```
Comment (georyetti) [inline src/kamae/spark/pipeline/pipeline.py]: Sorry spotted we don't use the reader here too. I know its not your change but could you use here too? `KamaeDefaultParamsReader`

**Review by georyetti — CHANGES_REQUESTED** [review]
Just one minor nitpick, we have kamae classes for reading and writing metadata/params, could we use them. We added these due to slowdowns in how metadata get written on databricks

**Review by georyetti — APPROVED** [review]

File: src/kamae/spark/pipeline/pipeline.py [hunk 14]
```diff
@@ -162,19 +531,258 @@ def _fit(self, dataset: DataFrame) -> "KamaeSparkPipelineModel":
...
+                )
+                model = stage.fit(fit_dataset)
                 transformers.append(model)
                 if stage in estimator_parent_stages:
-                    dataset = model.transform(dataset)
+                    full_dataset = model.transform(full_dataset)
+                    if sample_dataset is not None:
+                        sample_dataset = model.transform(sample_dataset)
+        if persisted_frame is not None:
+            persisted_frame.unpersist()
```
Comment (Sbranikas) [inline src/kamae/spark/pipeline/pipeline.py:785]: Just a small question, but right now only sampled_dataset gets cleaned up on failure (via the outer _fit's finally) — persisted_frame doesn't, right?

**Review by Sbranikas — APPROVED** [review]
Just a minor question on the unpersist().
